packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,371 @@
+      const plugin = this;
+      return function patchOpen(
+        this: XMLHttpRequest,
+        method: string,
+        url: string,
+        async?: boolean,
+        user?: string | null,
+        pass?: string | null
+      ): OpenFunction | any {
+        plugin._createSpan(this as XMLHttpRequestWrapped, url, method);
+
+        if (async) {
+          return original.call(this, method, url, async, user, pass);
+        } else {
+          return original.call(this, method, url, true, null, null);
+        }
+      };
+    };
+  }
+
+  /**
+   * Patches the method send
+   * @private
+   */
+  protected _patchSend() {
+    const plugin = this;
+
+    function endSpan(xhr: XMLHttpRequestWrapped, eventName: string) {
+      const spanId = xhr.__OT_SPAN_ID;
+      const callbackToRemoveEvents = spanId
+        ? plugin._callbackToRemoveEvents[spanId]
+        : undefined;
+
+      if (typeof callbackToRemoveEvents === 'function') {
+        callbackToRemoveEvents();
+      }
+      const currentSpan = spanId ? plugin._spans[spanId] : undefined;
+      if (currentSpan) {
+        plugin._addSpanNetworkEvents(currentSpan);
+        currentSpan.addEvent(eventName);
+
+        const url = currentSpan.attributes[AttributeNames.HTTP_URL];
+        if (typeof url === 'string') {
+          const parsedUrl = parseUrl(url);
+
+          currentSpan.setAttribute(AttributeNames.HTTP_STATUS_CODE, xhr.status);
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_STATUS_TEXT,
+            xhr.statusText
+          );
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_HOSTNAME,
+            parsedUrl.hostname
+          );
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_PATH,
+            parsedUrl.pathname
+          );
+
+          // @TODO do we want to collect this or it will be collected earlier once only or
+          //    maybe when parent span is not available ?
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_USER_AGENT,
+            navigator.userAgent
+          );
+        }
+
+        currentSpan.end();
+      }
+    }
+
+    function onError(this: XMLHttpRequestWrapped) {
+      endSpan(this, EventNames.EVENT_ERROR);
+    }
+
+    function onTimeout(this: XMLHttpRequestWrapped) {
+      endSpan(this, EventNames.EVENT_TIMEOUT);
+    }
+
+    function onLoad(this: XMLHttpRequestWrapped) {
+      if (this.status < 299) {
+        endSpan(this, EventNames.EVENT_LOAD);
+      } else {
+        endSpan(this, EventNames.EVENT_ERROR);
+      }
+    }
+
+    function unregister(xhr: XMLHttpRequestWrapped, spanId: string) {
+      xhr.removeEventListener('load', onLoad);
+      xhr.removeEventListener('error', onError);
+      xhr.removeEventListener('timeout', onTimeout);
+      delete plugin._callbackToRemoveEvents[spanId];
+      delete xhr.__OT_SPAN_ID;
+    }
+
+    return (original: SendFunction): SendFunction => {
```

Comment:
it looks like this already expects a span to be created? If so, can you mention in the jsdoc where the span is expected to be created and why this function does not create it?

Reply:
Where do you think it should be mentioned exactly ?
Some background:
You can't call `Send` method without calling `Open` method first which is the beginning of XMLHttpRequest. That's why the span is being created in `Open`. Also you don't have access to the `url` or `method` in `Send` function so it is straightforward the span is being created in the beginning.

Reply:
When i was originally reading it I didn't realize you created the span in 'open' and closed it in 'send' so this was just a misunderstanding on my part.